scripts/ubiquiti-discovery.nse

```diff
@@ -0,0 +1,254 @@
+  pos = 1
+
+  while pos <= #config_data - 2 do
+    tlv_type = config_data:byte(pos)
+    tlv_len  = string.unpack(">I2", config_data, pos +1)
+    pos = pos + 3
+
+    -- Sanity check that TLV len isn't larger than the data we have left.
+    -- Has been observed in the wild against protocols just similar enough to
+    -- make it here.
+    if tlv_len > (#config_data - pos + 1) then
+      return nil
+    end
+
+    tlv_value = config_data:sub(pos, pos + tlv_len - 1)
+
+    -- MAC address
+    if tlv_type == 0x01 then
+      mac_raw = tlv_value:sub(1, 6)
+      mac = stdnse.format_mac(mac_raw)
+      unique_macs[mac] = true
+
+    -- MAC and IP address
+    elseif tlv_type == 0x02 then
+      mac_raw = tlv_value:sub(1, 6)
+      mac = stdnse.format_mac(mac_raw)
+      unique_macs[mac] = true
+
+      ip_raw = tlv_value:sub(7, tlv_len)
+      ip = ipOps.str_to_ip(ip_raw)
+      mac_ip_table[mac] = ip
+
+    elseif tlv_type == 0x03 then
+      info.firmware = tlv_value
+
+      local human_version = tlv_value:match("%.(v%d+%.%d+%.%d+)")
+      if human_version then
+        info.version = human_version
+      end
+
+    elseif tlv_type == 0x0a then
+      local uptime_raw = string.unpack(">I4", tlv_value)
+      info.uptime_seconds = uptime_raw
+      info.uptime = uptime_str(uptime_raw)
+
+    elseif tlv_type == 0x0b then
+      info.hostname = tlv_value
+
+    elseif tlv_type == 0x0c then
+      info.product = tlv_value
+
+    elseif tlv_type == 0x0d then
+      info.essid = tlv_value
+
+    elseif tlv_type == 0x0f then
+      -- value also includes bit shifted flag for http vs https but we
+      -- are ignoring it here.
+      if tlv_len == 4 then
+        tlv_value = string.unpack(">I4", tlv_value)
+        info.mgmt_port = tlv_value & 0xffff
+      end
+
+    elseif tlv_type == 0x14 then
+      info.model = tlv_value
+
+    else
+
+    -- Other known or observed values
+    -- Some have been seen in code but not observed to test while others have
+    -- been observed but we don't know how to decode them.
+
+    -- 0x06 - username
+    -- 0x07 - salt
+    -- 0x08 - random challenge
+    -- 0x09 - challenge
+    -- 0x0e - WMODE - state of config? length 1 value 03 value 02
+    -- 0x10 - length 2 value e4b2 value e8a5 e815
+    -- 0x18 - length 4 and 4 nulls, or length 1 and 0xff
+    -- 0xff - length 2 value e835
+
+      stdnse.debug1("Unknown tag: %s - length: %d value: %s",
+                    stdnse.tohex(tlv_type), tlv_len,
+                    stdnse.tohex(tlv_value))
+    end
+
+    pos = pos + tlv_len
+  end
+
+  if mac_ip_table then
+    info.mac_ip = mac_ip_table
+  end
+
+  if unique_macs then
+    info.mac_addresses = {}
+    for k, _ in pairs(unique_macs) do
+      table.insert(info.mac_addresses, k)
+    end
+  end
+
+  return info
+end
+
+function action(host, port)
+
+  local socket = nmap.new_socket()
+  socket:connect(host, port)
+  socket:send(PROBE_V1)
+
+  local status, response = socket:receive()
```

Comment:
My understanding is that `socket:receive` is non-blocking so this might be subject to race. Would it make sense to use `socket:receive_bytes(4)` instead?

Also, should this be followed with `socket:close()`?

Reply:
Good call.